npm/pkg/dataplane/policies/chain-management_linux.go

```diff
@@ -294,6 +243,235 @@ func (pMgr *PolicyManager) bootup(_ []string) error {
 	return nil
 }
 
+// detectIptablesVersion sets the global iptables variable to nft if detected or legacy if detected.
+// NPM will crash if it fails to detect either.
+// This global variable is referenced in all iptables related functions.
+func (pMgr *PolicyManager) detectIptablesVersion() error {
+	klog.Info("first attempt detecting iptables version. running: iptables-nft-save -t mangle")
+	cmd := pMgr.ioShim.Exec.Command(util.IptablesSaveNft, "-t", "mangle")
+	output, err := cmd.CombinedOutput()
+	if err == nil && strings.Contains(string(output), "KUBE-IPTABLES-HINT") || strings.Contains(string(output), "KUBE-KUBELET-CANARY") {
+		msg := "detected iptables version on first attempt. found KUBE chains in nft iptables. NPM will use iptables-nft"
+		klog.Info(msg)
+		metrics.SendLog(util.IptmID, msg, metrics.DonotPrint)
+		util.SetIptablesToNft()
+		return nil
+	}
+
+	if err != nil {
+		msg := "failed to detect iptables version on first attempt. error running iptables-nft-save. will try detecting using iptables-legacy-save. err: %s"
+		metrics.SendErrorLogAndMetric(util.IptmID, msg, err.Error())
+	}
+
+	klog.Info("second attempt detecting iptables version. running: iptables-legacy-save -t mangle")
+	lCmd := pMgr.ioShim.Exec.Command(util.IptablesSaveLegacy, "-t", "mangle")
+	loutput, err := lCmd.CombinedOutput()
+	if err == nil && strings.Contains(string(loutput), "KUBE-IPTABLES-HINT") || strings.Contains(string(loutput), "KUBE-KUBELET-CANARY") {
```

Comment:
What is err is nil and string does not contain HINT chain, can we default to NFT here ?

Reply:
Update per offline discussion

#### Scenario 1
1. nft-save fails (e.g. seg fault)
2. legacy tables works but no hint

default to nft even if kernel version is lower, 

#### Scenario 2
1. nft-save works but no hint
2. legacy works but no hint

tie to kernal version

Reply:
New logic:

> Now we update the logic based on [this article](https://kubernetes.io/blog/2022/09/07/iptables-chains-not-api/#use-case-iptables-mode) to check for either kube chain in NFT. If either is there, use nft. If it's not there or there is any failure, check for either kube chain in legacy. If found, use legacy. Otherwise, check the kernel version. If version is >=5, use nft.

Reply:
New logic:

> check for either kube chain in NFT. If either is there, use nft. If it's not there or there is any failure, check for either kube chain in legacy. If found, use legacy. Otherwise, default to nft.